src/hooks/api/useFetch.js

```diff
@@ -0,0 +1,127 @@
+import { useState, useEffect, useRef } from "react";
+import axios from "axios";
+import { baseApiUrl } from "@/config/environment";
+import { logInfo } from "@/utils/logging";
+
+const useFetch = (initialRoute, onReceived) => {
+  // Validate initial inputs to avoid confusion with server routing
+  // The route cannot include 'api/' as part of the endpoint, to prevent conflicting or ambiguous routes like "api/api/log-in"
+  if (!initialRoute || initialRoute.includes("api/")) {
+    throw new Error(
+      "Invalid route provided. Routes cannot include 'api/' as part of the endpoint, to avoid conflicts and confusion in server routing.",
+    );
+  }
+
+  if (typeof initialRoute !== "string") {
+    throw new Error("useFetch: route must be a string");
+  }
+
+  if (typeof onReceived !== "function") {
+    throw new Error("useFetch: onReceived must be a function");
+  }
+
+  const [error, setError] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [route, setRoute] = useState(initialRoute);
+  const [data, setData] = useState(null);
+
+  // Store multiple cancel tokens
+  const cancelTokens = useRef({});
+
+  const performFetch = async (options = {}, newUrl) => {
+    if (newUrl) {
+      cancelFetch(newUrl); // Cancel the previous request if URL changes
+      setRoute(newUrl);
+    }
+
+    setError(null);
+    setData(null);
+    setIsLoading(true);
+
+    // Validate the route format
+    if (!route || !/^\/[a-zA-Z0-9/_-]*(\?[a-zA-Z0-9=&]*)?$/.test(route)) {
+      setError(new Error("Invalid URL"));
+      setIsLoading(false);
+      return;
+    }
+
+    const baseOptions = {
```

Comment:
Overall, Really great work with all the detailed checks before sending requests! You made the hook super robust and thoughtful! Just wondering, is this hook supposed to handle other request types like POST, UPDATE, DELETE, etc. or support custom headers and options to make it reusable?

Reply:
---

You’re absolutely right—this is something I hadn’t fully considered. The hook is designed so the calling component specifies the method (POST, DELETE, PUT, etc.) and any special headers or configurations as needed. For example:

```javascript
performFetch({
  method: "POST",
  data: { user: credentials },
});
```

The `data` here corresponds to the body of the request. 

If you’d like, we could explore ways to extend the hook to handle methods more seamlessly. Would you be interested in tackling that? 😊

---

Reply:
I've already done with this part in my last PR of the useFetch, you can check it and maybe we can combine mine to yours:)

Reply:
Okey I'm going to work on this over the weekend to see if it's feasible to combine the approaches!

Reply:
> I've already done with this part in my last PR of the useFetch, you can check it and maybe we can combine mine to yours:)

Hi Youjiao,

I’ve made the suggested adjustments:

Our useFetch hook now accepts all types of HTTP requests.
I updated the test file to cover scenarios for these new HTTP methods.
Looking forward to your feedback.

Reply:
I like that the hook is now more versatile, but now the call expects these fields. So, even if not all fields are required, they still need to be passed in the call.

I liked the previous method where you could easily override the HTTP methods, but the advantage of the current approach is that it's clearer for a programmer who isn't familiar with the code to understand what is expected. We can keep it this way if you think it's better, and the good thing is that we have now tested all types of requests.

Reply:
Super robust enhancement Lenin! I'm so impressed with that! Only one thing that I don't understand is why we have  still need to pass `body` or `headers` which are not strictly required? I think in your code is super comprehensive, It has default `null` for body and default empty object `{}` for `customHeaders` so it totally ok to just call useFetch easily like this: 
``` javascript
const { isLoading, error, data, performFetch } = useFetch("/users");

Reply:
Maybe you're right, but when I was working on the test, I had to pass the body and headers in the request, otherwise the test would fail. :/

Reply:
Okay, haha I never knew that. Ok it's fine, let's keep it there. Really impressive btw:)

Reply:
I see your point, and you're right—there's always a trade-off between flexibility and simplicity. This version of the hook is more versatile, but the previous one was more concise. For me, I think this is a solid foundation that we can continue to improve and refine in future pull requests.